xapian-letor/ranker/listmle_ranker.cc

```diff
@@ -0,0 +1,211 @@
+
+#include <xapian.h>
+#include <xapian/intrusive_ptr.h>
+#include <xapian/types.h>
+#include <xapian/visibility.h>
+
+#include <algorithm>
+#include <cmath>
+#include <stdlib.h>
+
+using namespace std;
+using namespace Xapian;
+
+ListMLERanker::~ListMLERanker()
+{
+    LOGCALL_DTOR(API, "ListMLERanker");
+}
+
+static bool
+label_comparer(const FeatureVector& firstfv,
+	       const FeatureVector& secondfv)
+{
+    return firstfv.get_label() > secondfv.get_label();
+}
+
+static double
+calculate_inner_product(const vector<double>& parameters,
+			const vector<double>& feature_sets)
+{
+    double inner_product = 0.0;
+    for (size_t i = 0; i < parameters.size(); ++i) {
+	inner_product += parameters[i] * feature_sets[i];
+    }
+    return inner_product;
+}
+
+static vector<double>
+calculate_gradient(const vector<FeatureVector>& sorted_feature_vectors,
+		   const vector<double>& new_parameters)
+{
+    //need to be optimized,how to get the length of the features
+    vector<double> gradient(sorted_feature_vectors[0].get_fcount(), 0);
+
+    size_t list_length = sorted_feature_vectors.size();
+
+    vector<double> exponents;
+    double expsum = 0.0;
+
+    for (size_t i = 0; i < list_length; ++i) {
+	double temp = exp(calculate_inner_product(
+			      new_parameters,
+			      sorted_feature_vectors[i].get_fvals()));
+	exponents.push_back(temp);
+	expsum += temp;
+    }
+
+    for (size_t i = 0; i < list_length; ++i) {
+	vector<double> feature_sets = sorted_feature_vectors[i].get_fvals();
+	for (size_t j = 0; j < feature_sets.size() - 1; ++j) {
+	    gradient[j] += feature_sets[j] * exponents[i] / expsum;
+	}
+    }
+
+    vector<double> first_place_in_ground_truth_feature_sets =
+	    sorted_feature_vectors[0].get_fvals();
+
+    for (size_t i = 0; i < first_place_in_ground_truth_feature_sets.size() - 1;
+	 ++i) {
+	gradient[i] -= first_place_in_ground_truth_feature_sets[i];
+    }
+
+    return gradient;
+}
+
+static void
+update_parameters(vector<double>& new_parameters,
+		  const vector<double>& gradient,
+		  const double& learning_rate)
+{
+    size_t num = new_parameters.size();
+    for (size_t i = 0; i < num; ++i) {
+	new_parameters[i] -= gradient[i] * learning_rate;
+    }
+}
+
+/* Updates new_parameters after calculating the gradient.
+ * new_parameters(w) is updated as: w = w - gradient * learningRate
+ */
+static void
+batch_learning(vector<FeatureVector> feature_vectors,
+	       vector<double> & new_parameters,
+	       const double& learning_rate)
+{
+    sort(feature_vectors.begin(), feature_vectors.end(), label_comparer);
+    vector<double> gradient = calculate_gradient(feature_vectors,
+						new_parameters);
+    update_parameters(new_parameters, gradient, learning_rate);
+}
+
+void
+ListMLERanker::train(const vector<FeatureVector>& training_data)
+{
+    LOGCALL_VOID(API, "ListMLERanker::train", training_data);
+    if (training_data.empty()) {
+	throw Xapian::InvalidArgumentError("Training data is empty."
```

Comment:
There needs to be a space between the two sentences in the final string. However, I don't think we need to call out the training _file_ here (because in future we might train from something other than a file — this class only knows it was given an empty training data vector). So maybe just something like `"Cannot train: no training data"` would be sufficient.